File: Level-2/29.java
class Solution {
    public int solution(int m, int n, String[] board) {
        int answer = 0;
        char[][] map = new char[m][n];
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                map[i][j] = board[i].charAt(j);
            }
        }
        
        boolean flag = true;
        while(flag){
            flag = false;
            int[][] visited = new int[m][n];
            for(int i = 0; i < m - 1; i++){
                for(int j = 0; j < n - 1; j++){
                    if(map[i][j] == '0') continue;
                    if(map[i][j] == map[i + 1][j] && map[i][j] == map[i][j + 1] && map[i][j] == map[i + 1][j + 1]){
                        visited[i][j] = 1;
                        visited[i + 1][j] = 1;
                        visited[i][j + 1] = 1;
                        visited[i + 1][j + 1] = 1;
                        
                        flag = true;
                    }
                }
            }
            
            for(int j = 0; j < n; j++){
                for(int i = 0; i < m; i++){
                    if(visited[i][j] == 1){
                        answer++;
                        for(int k = i - 1; k >= 0; k--){
                            map[k + 1][j] = map[k][j];
                            map[k][j] = '0';
                        }
                    }
                }
            }
        }
        
        return answer;
    }
}

File: Level-2/29.js
function solution(m, musicinfos) {
    let answer = [];
    m = filterString(m);
    
    for(let i = 0; i < musicinfos.length; i++) {
        let temp = musicinfos[i].split(',');
        let [start, end, title, code] = temp;
        code = filterString(code);
        
        let play_time = countTime(start, end);
        
        if(play_time < code.length) {
            code = code.slice(0, play_time);
        }else {
            code = code.repeat(Math.ceil(play_time / code.length));
        }
        
        if(code.includes(m)) {
            answer.push([play_time, title]);
        }
    }
    
    answer.sort((a, b) => b[0] - a[0]);
    
    return answer.length === 0 ? '(None)' : answer[0][1];
}

function filterString(string) {
    let result = string.replace(/(C#)/g, 'c')
                       .replace(/(D#)/g, 'd')
                       .replace(/(F#)/g, 'f')
                       .replace(/(G#)/g, 'g')
                       .replace(/(A#)/g, 'a');
    return result;
}

function countTime(start, end) {
    let [s_h, s_m] = start.split(':');
    let [e_h, e_m] = end.split(':');
    return (e_h - s_h) * 60 + (e_m - s_m);
}
